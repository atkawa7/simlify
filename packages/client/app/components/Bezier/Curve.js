import React from "react";

const Curve = (props) => {

  const {
    curveColor,
    curveWidth,
    positions,
  } = props;

  const [
    originX,
    originY,
    bezierPoint1X,
    bezierPoint1Y,
    bezierPoint2X,
    bezierPoint2Y,
    destinationX,
    destinationY,
  ] = positions;

  const curve = `M${originX},${originY}
    C${bezierPoint1X},${bezierPoint1Y}
    ${bezierPoint2X},${bezierPoint2Y}
    ${destinationX},${destinationY}`;

    return (
      <path
        fill="none"
        stroke={curveColor}
        strokeWidth={curveWidth}
        d={curve}
      />
    );
}

export default Curve;
