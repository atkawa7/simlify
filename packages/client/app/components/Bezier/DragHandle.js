import React, { useState } from 'react';

const DragHandle = (props) => {
  const {
    right,
    bottom,
    width,
    height,
    position = [0.2, 0.2],
  } = props;

  const dragHandleColor = '#00c0ff';
  const [hover, setHover] = useState(false);
  const [down, setDown] = useState(false);
  const handleStroke = 3;
  const handleRadius = 5;

  function relativeToAbsolutePos(relativePosition) {
    return [relativePosition[0] * width, height - relativePosition[1] * height];
  }

  let strokeOriginX = 0;
  let strokeOriginY = 0;
  let [strokeDestinationX, strokeDestinationY] = relativeToAbsolutePos(position);

  if(bottom) {
    strokeOriginY = height;
  }
  if(right) {
    strokeOriginX = width;
  }

  const onMouseEnter = () => setHover(true);
  const onMouseLeave = () => setHover(false);
  const onMouseUp = () => {
    setDown(false);
    props.onMouseUp();
  };
  const onMouseDown = () => {
    setDown(true);
    props.onMouseDown();
  };

  return(
    <g>
      <line
        stroke={dragHandleColor}
        strokeWidth={hover||down ? 1 + handleStroke : handleStroke}
        x1={strokeOriginX}
        y1={strokeOriginY}
        x2={strokeDestinationX}
        y2={strokeDestinationY} />
      <circle
        cx={strokeDestinationX}
        cy={strokeDestinationY}
        r={handleRadius}
        stroke={dragHandleColor}
        strokeWidth={hover||down ? 2 * handleStroke : handleStroke}
        fill={dragHandleColor}
        onMouseEnter={onMouseEnter}
        onMouseLeave={onMouseLeave}
        onMouseDown={onMouseDown}
        onMouseUp={onMouseUp}
      />
    </g>
  );
}

export default DragHandle;
