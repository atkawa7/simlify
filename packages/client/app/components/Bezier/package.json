{
    "name": "Bezier",
    "private": true,
    "main": "./Bezier.js"
}