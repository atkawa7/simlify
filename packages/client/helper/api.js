import axios from "axios";

const getRequest = async (url) => {
    const response = await axios.get(url);
    return response.data;
};

const postRequest = async (url, body) => {
    const response = await axios.post(url, body);
    return response.data;
};

const putRequest = async (url, body) => {
    const response = await axios.put(url, body);
    return response.data;
};

export default {
    getAvailableNodes: async () => await getRequest('http://localhost:3000/api/v1/nodes/list'),
    getFlows: async () => await getRequest('http://localhost:3000/api/v1/flows/all'),
    putFlow: async (updatedFlow) => await putRequest('http://localhost:3000/api/v1/flows', updatedFlow),
    postFlow: async (newFlow) => await postRequest('http://localhost:3000/api/v1/flows', newFlow),
};