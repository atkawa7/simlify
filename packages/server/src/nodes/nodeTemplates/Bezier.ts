'use strict';

import { NodeDataBase } from '../nodeBase';
import { OutputPort, InputPort } from '../ports';
import { portTypeFactory, timeSeriesPortType } from '../ports/portTypes';
import Flow from '../../flow/lib/Flow';
import { OptionsBase, VisualisationType } from '../nodeBase/NodeBase';

export default class Bezier extends NodeDataBase {
  value: number;

  constructor(parentFlow: Flow, nodeId: string) {
    super(parentFlow, nodeId);
    super.addPort(new InputPort(
      this,
      portTypeFactory.createNumberPortType(),
      'NumberOfPoints',
      200
    ));

    super.addPort(new OutputPort(
        this,
        portTypeFactory.createTimeSeriesPortType(),
        '',
        async () => await this.calculateBezierCurve()
    ));

    const options: OptionsBase = {
      variables: {
        points: [0.2, 0.2, 0.8, 0.8]
      },
      visualisation: VisualisationType.BezierCurve,
    };
    this.setOptions(options);
  }

  cubicBezier(p0: number, p1: number, p2: number, p3: number, t: number) {
    const term1 = p0 * Math.pow(1 - t, 3);
    const term2 = 3 * p1 * Math.pow(1 - t, 2) * t;
    const term3 = 3 * p2 * (1 - t) * Math.pow(t, 2);
    const term4 = p3 * Math.pow(t, 3);
    return  term1 + term2 + term3 + term4;
  }

  bezierCurve(points: number[] = [0.2, 0.2, 0.8, 0.8], tSteps: number) {
    const bezierXPoints = [];
    const bezierYPoints = [];
    for (let i = 0; i < tSteps; i = i + 1) {
      bezierXPoints.push(this.cubicBezier(0, points[0], points[2], 1, i / tSteps));
      bezierYPoints.push(this.cubicBezier(0, points[1], points[3], 1, i / tSteps));
    }
    return { bezierXPoints, bezierYPoints };
  }

  mapBezierPointsToArray(
    bezierPoints: { bezierXPoints: number[], bezierYPoints: number[] },
    numberOfPoints: number
  ) {
    const { bezierXPoints, bezierYPoints } = bezierPoints;
    let currentXPos = 0;
    const xStepSize = 1 / numberOfPoints;
    const bezierPointsLinear: { name: number, value: number }[] = [];

    bezierXPoints.forEach((xPoint, index) => {
      if (xPoint > currentXPos) {
        bezierPointsLinear.push({ name: xPoint, value: bezierYPoints[index] });
        currentXPos += xStepSize;
      }
    });

    return bezierPointsLinear;
  }

  async calculateBezierCurve() {
    const { NumberOfPoints } = await this.fetchInputPorts();
    const MIN_T_STEPS = 200;
    const xStepAmount = Math.max(NumberOfPoints * 10, MIN_T_STEPS);
    const bezierPoints = this.bezierCurve(this.getOptions().variables.points as any, xStepAmount);
    const bezierPointsLinear = this.mapBezierPointsToArray(bezierPoints, NumberOfPoints);
    return bezierPointsLinear;
  }
}
