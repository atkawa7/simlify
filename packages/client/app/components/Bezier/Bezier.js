import React, { useState } from 'react';
import Curve from './Curve';
import DragHandle from './DragHandle';

import './Bezier.scss';

const Bezier = (props) => {
  const {
    width,
    height,
    className = '',
    styles = {},
    defaultPosition = [0.2, 0.2, 0.8, 0.8],
    onChange,
  } = props;

  const [selectedHandler, setSelectedHandler] = useState(null);
  const [handler1Position, setHandler1Position] = useState([defaultPosition[0], defaultPosition[1]]);
  const [handler2Position, setHandler2Position] = useState([defaultPosition[2], defaultPosition[3]]);

  const ref = React.createRef();

  function absoluteToRelativePos(absolutePos) {
    const boundingBox = ref.current.getBoundingClientRect();
    const xRelative = (absolutePos[0] - boundingBox.left) / boundingBox.width;
    const yRelative = -(absolutePos[1] - boundingBox.top - boundingBox.height) / boundingBox.height;
    return [xRelative, yRelative];
  };

  function onMouseMove(event) {
    const relPos = absoluteToRelativePos([event.clientX, event.clientY]);
    let positionChanged = false;
    if(selectedHandler === 1) {
      setHandler1Position(relPos);
      positionChanged = true;
    }
    if(selectedHandler === 2) {
      setHandler2Position(relPos);
      positionChanged = true;
    }
    if(positionChanged && onChange) onChange(handler1Position.concat(handler2Position));
  }

  function onMouseUp(event) {
    
  }

  function onMouseLeave(event) {
    
  }

  function createCurvePointArray() {
    const relativePoints = [0, 0]
      .concat(handler1Position[0], handler1Position[1])
      .concat(handler2Position[0], handler2Position[1])
      .concat([1, 1]);
    return relativePoints.map((point, index) => index % 2 ? (1 - point) * width : point * height)
  }

  const setHandlerDown = (handlerNumber) => {
    setSelectedHandler(handlerNumber);
  }

  const mouseEvents = {
    onMouseMove,
    onMouseUp,
    onMouseLeave,
  };

  return (
    <svg
      ref={ref}
      className={`bezier__background ${className}`}
      style={styles}
      width={width}
      height={height}
      {...mouseEvents}
    >
      <Curve
        curveColor="#aaaaaa"
        curveWidth={3}
        positions={createCurvePointArray()}
      />
      <DragHandle 
        width={width}
        height={height}
        position={handler1Position}
        onMouseUp={() => setHandlerDown(null)}
        onMouseDown={() => setHandlerDown(1)}
        bottom
      />
      <DragHandle 
        width={width}
        height={height}
        position={handler2Position}
        onMouseUp={() => setHandlerDown(null)}
        onMouseDown={() => setHandlerDown(2)}
        right
      />
    </svg>
  );
};

export default Bezier;
