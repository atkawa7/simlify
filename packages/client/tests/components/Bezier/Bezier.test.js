import React from 'react';
import api from '../../../helper/api';
import Bezier from '../../../app/components/Bezier';
import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const snapshot = renderer
    .create(<Bezier />)
    .toJSON();
  expect(snapshot).toMatchSnapshot();
});