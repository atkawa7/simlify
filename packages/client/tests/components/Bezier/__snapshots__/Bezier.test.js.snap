// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<svg
  className="bezier__background "
  onMouseLeave={[Function]}
  onMouseMove={[Function]}
  onMouseUp={[Function]}
  style={Object {}}
>
  <path
    d="MNaN,NaN
    CNaN,NaN
    NaN,NaN
    NaN,NaN"
    fill="none"
    stroke="#aaaaaa"
    strokeWidth={3}
  />
  <g>
    <line
      stroke="#00c0ff"
      strokeWidth={3}
      x1={0}
      x2={NaN}
      y2={NaN}
    />
    <circle
      cx={NaN}
      cy={NaN}
      fill="#00c0ff"
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      r={5}
      stroke="#00c0ff"
      strokeWidth={3}
    />
  </g>
  <g>
    <line
      stroke="#00c0ff"
      strokeWidth={3}
      x2={NaN}
      y1={0}
      y2={NaN}
    />
    <circle
      cx={NaN}
      cy={NaN}
      fill="#00c0ff"
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      r={5}
      stroke="#00c0ff"
      strokeWidth={3}
    />
  </g>
</svg>
`;
